### PR TITLE
[DEV-1721] Enforce time limits for periodic tasks scheduled based on intervals

### PR DESCRIPTION
**Pull request by kchua78**
## Description

Periodic tasks scheduled based on intervals can pile up if such task does not complete within the interval.

Changes made in this PR:

- Always specify soft time limit for scheduled tasks (defaults to interval length), can be overriden
- Support enforcement of soft time limit for io-intensive tasks executed in gevent workers (soft limit is not supported by celery for gevent worker pool)

## Related Issue

https://featurebyte.atlassian.net/browse/DEV-1721

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [x] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [x] I have written tests for the changes made.
- [x] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [x] I have labeled my Pull Request correctly

### CODE REVIEW

File: featurebyte/worker/task_executor.py [hunk 1]
```diff
@@ -61,13 +69,20 @@ def run_async(coro: Awaitable[Any]) -> Any:
...
+    try:
+        with gevent.Timeout(seconds=timeout, exception=TimeoutError):
+            event = gevent.event.Event()
+            future.add_done_callback(lambda _: event.set())
+            event.wait()
+            return future.result()
+    except TimeoutError as exc:
+        # try to cancel the job if it has not started
+        future.cancel()
+        raise SoftTimeLimitExceeded(f"Task timed out after {timeout}s") from exc
```
Comment (gee-senbong) [inline featurebyte/worker/task_executor.py:85]: Will it throw another exception (when calling `future.cancel()`) if the job is started already?

File: featurebyte/worker/task_executor.py [hunk 2]
```diff
@@ -61,13 +69,20 @@ def run_async(coro: Awaitable[Any]) -> Any:
...
-    event = gevent.event.Event()
-    future.add_done_callback(lambda _: event.set())
-    event.wait()
-    return future.result()
+    try:
+        with gevent.Timeout(seconds=timeout, exception=TimeoutError):
+            event = gevent.event.Event()
+            future.add_done_callback(lambda _: event.set())
+            event.wait()
+            return future.result()
```
Comment (gee-senbong) [inline featurebyte/worker/task_executor.py:81]: Seems like we could pass the `timeout` parameter to `future.result()` here (https://docs.python.org/3/library/asyncio-task.html#scheduling-from-other-threads).

File: featurebyte/worker/task_executor.py (hunk 1 above)
Comment (kchua78) [reply featurebyte/worker/task_executor.py:85]: cancel will not raise any exception but return False if the task cannot be cancelled

File: featurebyte/worker/task_executor.py (hunk 2 above)
Comment (kchua78) [reply featurebyte/worker/task_executor.py:81]: Tried that - it raises timeout error but unfortunately the async task continues to run in the asyncio thread

**Review by gee-senbong — APPROVED** [review]

**Pull request merged**